include/zephyr/modem/ubx.h

```diff
@@ -0,0 +1,138 @@
+
+	struct modem_pipe *pipe;
+
+	struct k_work send_work;
+	struct k_work process_work;
+	struct k_sem script_stopped_sem;
+	struct k_sem script_running_sem;
+};
+
+struct modem_ubx_config {
+	void *user_data;
+
+	uint8_t *receive_buf;
+	uint16_t receive_buf_size;
+	uint8_t *work_buf;
+	uint16_t work_buf_size;
+	uint8_t *ubx_response_buf;
+	uint16_t ubx_response_buf_size;
+
+	k_timeout_t process_timeout;
+};
+
+/**
+ * @brief Attach pipe to Modem Ubx
+ *
+ * @param ubx Modem Ubx instance
+ * @param pipe Pipe instance to attach Modem Ubx instance to
+ * @returns 0 if successful
+ * @returns negative errno code if failure
+ * @note Modem Ubx instance is enabled if successful
+ */
+int modem_ubx_attach(struct modem_ubx *ubx, struct modem_pipe *pipe);
+
+/**
+ * @brief Release pipe from Modem Ubx instance
+ *
+ * @param ubx Modem Ubx instance
+ */
+void modem_ubx_release(struct modem_ubx *ubx);
+
+/**
+ * @brief Initialize Modem Ubx instance
+ * @param ubx Modem Ubx instance
+ * @param config Configuration which shall be applied to the Modem Ubx instance
+ * @note Modem Ubx instance must be attached to a pipe instance
+ */
+int modem_ubx_init(struct modem_ubx *ubx, const struct modem_ubx_config *config);
+
+/**
+ * @brief Writes the ubx frame in the script and reads it's response
+ * @details For each ubx frame sent, the device responds with a UBX-ACK frame (class = 0x05).
+ *	In case a "get" or "poll" ubx frame (used to get configuration of the device) is sent,
+ *	the device responds with two back-to-back frames - first with the exact same ubx frame
+ *	whose payload contains the desired configuration; then sends the UBX-ACK frame.
+ *	In case a "set" ubx frame (used to set configuration of the device) is sent, the device
+ *	only responds with a UBX-ACK frame.
+ *	Ex: if "set" variant of UBX-CFG-GNSS was sent, then the device responds with UBX-ACK only.
+ *	Ex: if "get" variant of UBX-CFG-GNSS was sent, then the device responds with UBX-CFG-GNSS
+ *	then sends a UBX-ACK message.
+ *
+ *	The message id of the UBX-ACK received determines whether the device acknowledged or not
+ *	acknowledged the ubx frame that was sent to it.
+ *
+ *	This function writes the ubx frame in the script and reads it's response from the device.
+ *	If a "get" response was received, then it's written back to the ubx_frame of the script,
+ *	so that the caller could retrieve the information that it required.
+ * @param ubx Modem Ubx instance
+ * @param script Script (containing the ubx frame) to be executed
+ * @note The length of ubx frame in the script should not exceed UBX_FRM_SZ_MAX
+ * @note Modem Ubx instance must be attached to a pipe instance
+ * @returns 0 if device acknowledged via UBX-ACK and no "get" response was received
+ * @returns positive integer denoting the length of "get" response that was received
+ * @returns negative errno code if failure
+ */
+int modem_ubx_run_script(struct modem_ubx *ubx, const struct modem_ubx_script *script);
+
```

Comment:
I believe there is an unclear separation between the responsibility of `ubx.h` and `gnss_u_blox_protocol.h` I understand that the GNSS specific macros and values like `UBX_CFG_GNSS_FLAG_SGN_CNF_GALILEO_E1` are in `gnss_u_blox_protocol.h`, but not `UBX_FRM_MSG_CLASS_IDX` and `UBX_PREAMBLE_SYNC_CHAR_1`. The latter are related to the UBX protocol itself, as opposed to the contents of the payload, which I would understand being split into its own header :)

As I understand it, `ubx.h` should not include `gnss_u_blox_protocol.h`, rather, everything it needs that is currently in `gnss_u_blox_protocol.h` should be moved to `ubx.h`, and `gnss_u_blox_protocol.h` should include `ubx.h` :)

With this, everything related to the UBX frame itself, is in `ubx.h`, checksums, flags, transport etc. while the different message classes, like `CFG(0x06)` are in a "child" header

Reply:
Understood. Just one question - `modem_ubx` needs the message class and message IDs of UBX-ACK message. And all messages classes and messages IDs are given in `gnss_u_blox_protocol_defines.h`. For now, to resolve the issue, I have added the following macros in `ubx.h`, but it causes some duplication:

```
#define UBX_MSG_CLASS_ACK	0x05
#define UBX_ACK_MSG_ID_ACK	0x01
#define UBX_ACK_MSG_ID_NAK	0x00
```
(These are GNSS specific macros.)

Reply:
We have a similar scenario for NMEA and AT messages :) The `modem_chat` module defines two things for each script, what to send, and what to expect to get back. When sending `AT+CEREG`, we expect to get `+CEREG=1,2` back. The `modem_chat` module has no concept of what any AT message is, or what should be returned, just like `modem_ubx` should not need to "understand" what the sent payload is, or what is expected to be returned.

A solution to this issue for `modem_ubx` is to write the script like this:

```
struct modem_ubx_frame {
        /* Class of UBX frame */
        uint8_t class;
        /* ID of UBX frame */
        uint8_t id;
        /* Source of payload for request
         * Destination for payload for response
         */
        uint8_t *payload;
        /* Size of payload to send for request
         * Expected size of payload for response
         */
        uint16_t payload_size;
};

struct modem_ubx_script {
        /* Pointer to request to send */
        struct modem_ubx_frame *request;
        /* Pointer to array of responses in the order they will be received */
        struct modem_ubx_frame *responses;
        /* Size of array of responses */
        uint8_t responses_size;
	uint16_t retry_count;
	k_timeout_t script_timeout;
};
```

This solution splits the encoding of a frame, including header and checksum, to `modem_ubx`, while `gnss_u_blox_protocol_defines.h` sets the classes and ids, formats the payload, and describes which frames are expected to be sent/received in which order :)

Reply:
I created this example of how to use it, converting this
```
	UBX_CFG_PRT_SET_DATA_INIT(frame_data)
	frame_data.baudrate = target_baudrate;

	struct modem_ubx_script script;

	ret = ubx_m10_modem_ubx_script_init(dev, &script, data->ubx_frame_buf,
		sizeof(data->ubx_frame_buf), retry, &frame_data, UBX_CLASS_CFG, UBX_CFG_PRT,
		UBX_CFG_PRT_SET_PAYLOAD_SZ);
	if (ret < 0) {
		goto out;
	}

	script.ubx_frame_size = ret;
	ret = ubx_m10_modem_ubx_run_script(dev, &script);
	if (ret < 0) {
		goto out;
	}
```
into
```
	UBX_CFG_PRT_SET_DATA_INIT(frame_data)
	frame_data.baudrate = target_baudrate;

	struct modem_ubx_frame cfg_prt_frame = {
		.class = UBX_CLASS_CFG,
		.id = UBX_CFG_PRT,
		.payload = (const uint8_t *)frame_data,
		.payload = UBX_CFG_PRT_SET_PAYLOAD_SZ,
	};

	struct modem_ubx_script cfg_prt_script = {
		.request = &cfg_prt_frame,
		.responses = NULL,
		.responses_size = 0,
		.retry_count = 3,
		.script_timeout = K_SECONDS(2),
	};

	ret = ubx_m10_modem_ubx_run_script(dev, &script);
	if (ret < 0) {
		goto out;
	}

```

Reply:
Thank you for answering the question @bjarki-trackunit. Really appreciate your help and effort.
I just pushed my latest changes as you could see below. Will address this comment as well and push the new changes shortly.

Reply:
Nice, I will review your latest input shortly